File: tools/cui/lexer.py
"""
lexer.py - Tokenizer for .cui DSL files.

String interpolation is handled by the lexer: a STRING token carries a
pre-parsed list of segments (StringSegment / InterpSegment).
"""

from __future__ import annotations
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional
from ast_nodes import StringLit, StringSegment, InterpSegment


class TT(Enum):
    # Keywords
    KW_VIEW    = auto()
    KW_LET     = auto()
    KW_FUNC    = auto()
    KW_ONCE    = auto()
    KW_ALWAYS  = auto()
    KW_VAR     = auto()
    KW_IMPORT  = auto()
    KW_TRUE    = auto()
    KW_FALSE   = auto()
    KW_RETURN  = auto()
    # Property wrappers
    AT_OBSERVED = auto()
    AT_MUTABLE  = auto()   # non-const writable reference
    AT_SNAPSHOT = auto()
    AT_STATE    = auto()
    AT_BINDING  = auto()
    AT_CUI_CPP  = auto()
    # Literals
    IDENT      = auto()
    INT_LIT    = auto()
    FLOAT_LIT  = auto()
    STRING     = auto()  # value is a StringLit
    # Operators / punctuation
    ARROW      = auto()  # ->
    EQ         = auto()  # =
    COLON      = auto()  # :
    COMMA      = auto()  # ,
    DOT        = auto()  # .
    DCOLON     = auto()  # ::
    LPAREN     = auto()  # (
    RPAREN     = auto()  # )
    LBRACE     = auto()  # {
    RBRACE     = auto()  # }
    QUESTION   = auto()  # ?
    PLUS       = auto()
    MINUS      = auto()
    STAR       = auto()
    SLASH      = auto()
    SEMICOLON  = auto()  # ; (block-form func)
    EOF        = auto()


KEYWORDS = {
    "view":    TT.KW_VIEW,
    "let":     TT.KW_LET,
    "func":    TT.KW_FUNC,
    "once":    TT.KW_ONCE,
    "always":  TT.KW_ALWAYS,
    "var":     TT.KW_VAR,
    "import":  TT.KW_IMPORT,
    "true":    TT.KW_TRUE,
    "false":   TT.KW_FALSE,
    "return":  TT.KW_RETURN,
}

WRAPPERS = {
    "@Observed": TT.AT_OBSERVED,
    "@Mutable":  TT.AT_MUTABLE,
    "@Snapshot": TT.AT_SNAPSHOT,
    "@State":    TT.AT_STATE,
    "@Binding":  TT.AT_BINDING,
    "@cui-cpp":  TT.AT_CUI_CPP,
}


@dataclass
class Token:
    type:  TT
    value: object   # str | float | int | StringLit | None
    line:  int = 0
    col:   int = 0

    def __repr__(self):
        return f"Token({self.type.name}, {self.value!r}, {self.line}:{self.col})"


class LexError(Exception):
    def __init__(self, msg, line, col):
        super().__init__(f"Lex error at {line}:{col}: {msg}")
        self.line = line
        self.col  = col


class Lexer:
    def __init__(self, source: str, filename: str = "<cui>"):
        self._src      = source
        self._pos      = 0
        self._line     = 1
        self._col      = 1
        self._filename = filename

    # ── public API ────────────────────────────────────────────────────────────

    def tokenize(self) -> list[Token]:
        tokens = []
        while True:
            tok = self._next_token()
            tokens.append(tok)
            if tok.type is TT.EOF:
                break
        return tokens

    # ── internals ─────────────────────────────────────────────────────────────

    def _peek(self, offset: int = 0) -> str:
        idx = self._pos + offset
        return self._src[idx] if idx < len(self._src) else "\0"

    def _advance(self) -> str:
        ch = self._src[self._pos]
        self._pos += 1
        if ch == "\n":
            self._line += 1
            self._col = 1
        else:
            self._col += 1
        return ch

    def _skip_whitespace_and_comments(self):
        while self._pos < len(self._src):
            ch = self._peek()
            if ch in " \t\r\n":
                self._advance()
            elif ch == "/" and self._peek(1) == "/":
                # Line comment
                while self._pos < len(self._src) and self._peek() != "\n":
                    self._advance()
            elif ch == "/" and self._peek(1) == "*":
                # Block comment
                self._advance(); self._advance()
                while self._pos < len(self._src):
                    if self._peek() == "*" and self._peek(1) == "/":
                        self._advance(); self._advance()
                        break
                    self._advance()
            else:
                break

    def _next_token(self) -> Token:
        self._skip_whitespace_and_comments()
        line, col = self._line, self._col

        if self._pos >= len(self._src):
            return Token(TT.EOF, None, line, col)

        ch = self._peek()

        # ── @ prefixed (wrappers / @cui-cpp) ──────────────────────────────
        if ch == "@":
            return self._lex_at(line, col)

        # ── String literal ─────────────────────────────────────────────────
        if ch == '"':
            return self._lex_string(line, col)

        # ── Numbers ────────────────────────────────────────────────────────
        if ch.isdigit() or (ch == "." and self._peek(1).isdigit()):
            return self._lex_number(line, col)

        # ── Identifiers / keywords ─────────────────────────────────────────
        if ch.isalpha() or ch == "_":
            return self._lex_ident(line, col)

        # ── Two-char operators ─────────────────────────────────────────────
        if ch == "-" and self._peek(1) == ">":
            self._advance(); self._advance()
            return Token(TT.ARROW, "->", line, col)

        if ch == ":" and self._peek(1) == ":":
            self._advance(); self._advance()
            return Token(TT.DCOLON, "::", line, col)

        # ── Single-char operators ──────────────────────────────────────────
        SINGLES = {
            "=": TT.EQ,    ":": TT.COLON,  ",": TT.COMMA,
            ".": TT.DOT,   "(": TT.LPAREN, ")": TT.RPAREN,
            "{": TT.LBRACE, "}": TT.RBRACE, "?": TT.QUESTION,
            "+": TT.PLUS,  "-": TT.MINUS,  "*": TT.STAR,
            "/": TT.SLASH, ";": TT.SEMICOLON,
        }
        if ch in SINGLES:
            self._advance()
            return Token(SINGLES[ch], ch, line, col)

        raise LexError(f"Unexpected character {ch!r}", line, col)

    # ── @ tokens ──────────────────────────────────────────────────────────────

    def _lex_at(self, line, col) -> Token:
        start = self._pos
        self._advance()  # consume '@'
        while self._pos < len(self._src) and (self._peek().isalnum() or self._peek() in "-_"):
            self._advance()
        word = self._src[start:self._pos]
        if word in WRAPPERS:
            return Token(WRAPPERS[word], word, line, col)
        raise LexError(f"Unknown @ token: {word!r}", line, col)

    # ── Identifiers ───────────────────────────────────────────────────────────

    def _lex_ident(self, line, col) -> Token:
        start = self._pos
        while self._pos < len(self._src) and (self._peek().isalnum() or self._peek() == "_"):
            self._advance()
        word = self._src[start:self._pos]
        tt = KEYWORDS.get(word, TT.IDENT)
        value = True if tt is TT.KW_TRUE else (False if tt is TT.KW_FALSE else word)
        return Token(tt, value, line, col)

    # ── Numbers ───────────────────────────────────────────────────────────────

    def _lex_number(self, line, col) -> Token:
        start = self._pos
        is_float = False
        while self._pos < len(self._src) and self._peek().isdigit():
            self._advance()
        if self._peek() == "." and self._peek(1).isdigit():
            is_float = True
            self._advance()
            while self._pos < len(self._src) and self._peek().isdigit():
                self._advance()
        # Scientific notation: 1e-6, 1.5E+3, etc.
        if self._peek() in ("e", "E"):
            next1 = self._peek(1)
            if next1.isdigit() or next1 in ("+", "-"):
                is_float = True
                self._advance()  # consume e/E
                if self._peek() in ("+", "-"):
                    self._advance()
                while self._pos < len(self._src) and self._peek().isdigit():
                    self._advance()
        # Optional suffix: .s .ms .us .ns (duration literals)
        if self._peek() == ".":
            suffix_start = self._pos
            self._advance()
            suffix = ""
            while self._pos < len(self._src) and self._peek().isalpha():
                suffix += self._advance()
            if suffix in ("s", "ms", "us", "ns"):
                raw = self._src[start:suffix_start]
                return Token(TT.FLOAT_LIT, _to_seconds(float(raw), suffix), line, col)
            else:
                # Not a duration suffix — put back the '.' and letters
                self._pos = suffix_start
        text = self._src[start:self._pos]
        if is_float or "." in text:
            return Token(TT.FLOAT_LIT, float(text), line, col)
        return Token(TT.INT_LIT, int(text), line, col)

    # ── String literals with interpolation ───────────────────────────────────

    def _lex_string(self, line, col) -> Token:
        self._advance()  # consume opening "
        segments = []
        buf = ""

        while self._pos < len(self._src):
            ch = self._peek()

            if ch == '"':
                self._advance()
                if buf:
                    segments.append(StringSegment(buf))
                return Token(TT.STRING, StringLit(segments), line, col)

            if ch == "\\":
                if self._peek(1) == "(":
                    # Interpolation
                    if buf:
                        segments.append(StringSegment(buf))
                        buf = ""
                    self._advance()  # '\'
                    self._advance()  # '('
                    capture, expr_tokens = self._lex_interp_expr()
                    segments.append(InterpSegment(capture, expr_tokens))
                    continue
                else:
                    # Simple escape
                    self._advance()
                    esc = self._advance()
                    ESC = {"n": "\n", "t": "\t", "r": "\r", '"': '"', "\\": "\\"}
                    buf += ESC.get(esc, esc)
                    continue

            buf += self._advance()

        raise LexError("Unterminated string literal", line, col)

    def _lex_interp_expr(self) -> tuple[Optional[str], list[Token]]:
        """Lex the expression inside \\(...). Returns (capture_mode, tokens)."""
        capture = None
        tokens = []
        depth = 1  # we already consumed the opening '('

        # Peek for once/always
        saved_pos  = self._pos
        saved_line = self._line
        saved_col  = self._col
        self._skip_whitespace_and_comments()
        if self._pos < len(self._src) and (self._peek().isalpha() or self._peek() == "_"):
            word_start = self._pos
            while self._pos < len(self._src) and (self._peek().isalnum() or self._peek() == "_"):
                self._advance()
            word = self._src[word_start:self._pos]
            if word in ("once", "always"):
                capture = word
            else:
                # Not a capture keyword — restore
                self._pos  = saved_pos
                self._line = saved_line
                self._col  = saved_col
        else:
            self._pos  = saved_pos
            self._line = saved_line
            self._col  = saved_col

        # Lex tokens until the matching ')'
        while self._pos < len(self._src):
            self._skip_whitespace_and_comments()
            if self._pos >= len(self._src):
                break
            ch = self._peek()
            if ch == "(" :
                depth += 1
                tok = self._next_token()
                tokens.append(tok)
            elif ch == ")":
                depth -= 1
                if depth == 0:
                    self._advance()  # consume ')'
                    break
                tok = self._next_token()
                tokens.append(tok)
            else:
                tok = self._next_token()
                tokens.append(tok)

        return capture, tokens


# ── Helpers ───────────────────────────────────────────────────────────────────

def _to_seconds(value: float, suffix: str) -> float:
    factors = {"s": 1.0, "ms": 1e-3, "us": 1e-6, "ns": 1e-9}
    return value * factors[suffix]

File: tools/cui-rs/src/lexer.rs
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TokenType {
    KwView,
    KwLet,
    KwFunc,
    KwOnce,
    KwAlways,
    KwVar,
    KwImport,
    KwTrue,
    KwFalse,
    KwReturn,
    AtObserved,
    AtMutable,
    AtSnapshot,
    AtState,
    AtBinding,
    AtCuiCpp,
    Ident,
    IntLit,
    FloatLit,
    String,
    Arrow,
    Eq,
    Colon,
    Comma,
    Dot,
    DColon,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Question,
    Plus,
    Minus,
    Star,
    Slash,
    Semicolon,
    Eof,
}

#[derive(Clone, Debug)]
pub enum StringTokenPart {
    Text(String),
    Interp {
        capture: Option<String>,
        tokens: Vec<Token>,
    },
}

#[derive(Clone, Debug)]
pub enum TokenValue {
    None,
    Text(String),
    Int(i64),
    Float(f64),
    Bool(bool),
    String(Vec<StringTokenPart>),
}

#[derive(Clone, Debug)]
pub struct Token {
    pub ty: TokenType,
    pub value: TokenValue,
    pub line: usize,
    pub col: usize,
}

impl Token {
    pub fn new(ty: TokenType, value: TokenValue, line: usize, col: usize) -> Self {
        Self {
            ty,
            value,
            line,
            col,
        }
    }

    pub fn text(&self) -> String {
        match &self.value {
            TokenValue::Text(value) => value.clone(),
            TokenValue::Int(value) => value.to_string(),
            TokenValue::Float(value) => format_float(*value),
            TokenValue::Bool(value) => value.to_string(),
            TokenValue::String(_) | TokenValue::None => String::new(),
        }
    }
}

#[derive(Clone, Debug)]
pub struct LexError {
    pub message: String,
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Lex error at {}:{}: {}",
            self.line, self.col, self.message
        )
    }
}

pub struct Lexer {
    chars: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
}

impl Lexer {
    pub fn new(source: &str) -> Self {
        Self {
            chars: source.chars().collect(),
            pos: 0,
            line: 1,
            col: 1,
        }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, LexError> {
        let mut tokens = Vec::new();
        loop {
            let token = self.next_token()?;
            let done = token.ty == TokenType::Eof;
            tokens.push(token);
            if done {
                break;
            }
        }
        Ok(tokens)
    }

    fn next_token(&mut self) -> Result<Token, LexError> {
        self.skip_whitespace_and_comments();
        let line = self.line;
        let col = self.col;

        if self.pos >= self.chars.len() {
            return Ok(Token::new(TokenType::Eof, TokenValue::None, line, col));
        }

        let ch = self.peek(0);
        if ch == '@' {
            return self.lex_at(line, col);
        }
        if ch == '"' {
            return self.lex_string(line, col);
        }
        if ch.is_ascii_digit() || (ch == '.' && self.peek(1).is_ascii_digit()) {
            return self.lex_number(line, col);
        }
        if is_ident_start(ch) {
            return Ok(self.lex_ident(line, col));
        }
        if ch == '-' && self.peek(1) == '>' {
            self.advance();
            self.advance();
            return Ok(Token::new(
                TokenType::Arrow,
                TokenValue::Text("->".to_string()),
                line,
                col,
            ));
        }
        if ch == ':' && self.peek(1) == ':' {
            self.advance();
            self.advance();
            return Ok(Token::new(
                TokenType::DColon,
                TokenValue::Text("::".to_string()),
                line,
                col,
            ));
        }

        let ty = match ch {
            '=' => TokenType::Eq,
            ':' => TokenType::Colon,
            ',' => TokenType::Comma,
            '.' => TokenType::Dot,
            '(' => TokenType::LParen,
            ')' => TokenType::RParen,
            '{' => TokenType::LBrace,
            '}' => TokenType::RBrace,
            '?' => TokenType::Question,
            '+' => TokenType::Plus,
            '-' => TokenType::Minus,
            '*' => TokenType::Star,
            '/' => TokenType::Slash,
            ';' => TokenType::Semicolon,
            other => {
                return Err(self.error_at(format!("Unexpected character {other:?}"), line, col));
            }
        };
        self.advance();
        Ok(Token::new(ty, TokenValue::Text(ch.to_string()), line, col))
    }

    fn lex_at(&mut self, line: usize, col: usize) -> Result<Token, LexError> {
        let start = self.pos;
        self.advance();
        while is_ident_continue(self.peek(0)) || self.peek(0) == '-' {
            self.advance();
        }
        let word: String = self.chars[start..self.pos].iter().collect();
        let ty = match word.as_str() {
            "@Observed" => TokenType::AtObserved,
            "@Mutable" => TokenType::AtMutable,
            "@Snapshot" => TokenType::AtSnapshot,
            "@State" => TokenType::AtState,
            "@Binding" => TokenType::AtBinding,
            "@cui-cpp" => TokenType::AtCuiCpp,
            _ => {
                return Err(self.error_at(format!("Unknown @ token: {word:?}"), line, col));
            }
        };
        Ok(Token::new(ty, TokenValue::Text(word), line, col))
    }

    fn lex_ident(&mut self, line: usize, col: usize) -> Token {
        let start = self.pos;
        while is_ident_continue(self.peek(0)) {
            self.advance();
        }
        let word: String = self.chars[start..self.pos].iter().collect();
        let (ty, value) = match word.as_str() {
            "view" => (TokenType::KwView, TokenValue::Text(word)),
            "let" => (TokenType::KwLet, TokenValue::Text(word)),
            "func" => (TokenType::KwFunc, TokenValue::Text(word)),
            "once" => (TokenType::KwOnce, TokenValue::Text(word)),
            "always" => (TokenType::KwAlways, TokenValue::Text(word)),
            "var" => (TokenType::KwVar, TokenValue::Text(word)),
            "import" => (TokenType::KwImport, TokenValue::Text(word)),
            "true" => (TokenType::KwTrue, TokenValue::Bool(true)),
            "false" => (TokenType::KwFalse, TokenValue::Bool(false)),
            "return" => (TokenType::KwReturn, TokenValue::Text(word)),
            _ => (TokenType::Ident, TokenValue::Text(word)),
        };
        Token::new(ty, value, line, col)
    }

    fn lex_number(&mut self, line: usize, col: usize) -> Result<Token, LexError> {
        let start = self.pos;
        let mut is_float = false;
        while self.peek(0).is_ascii_digit() {
            self.advance();
        }
        if self.peek(0) == '.' && self.peek(1).is_ascii_digit() {
            is_float = true;
            self.advance();
            while self.peek(0).is_ascii_digit() {
                self.advance();
            }
        }
        if matches!(self.peek(0), 'e' | 'E') {
            let next = self.peek(1);
            if next.is_ascii_digit() || matches!(next, '+' | '-') {
                is_float = true;
                self.advance();
                if matches!(self.peek(0), '+' | '-') {
                    self.advance();
                }
                while self.peek(0).is_ascii_digit() {
                    self.advance();
                }
            }
        }
        if self.peek(0) == '.' {
            let suffix_start = self.pos;
            self.advance();
            let mut suffix = String::new();
            while self.peek(0).is_ascii_alphabetic() {
                suffix.push(self.advance());
            }
            if matches!(suffix.as_str(), "s" | "ms" | "us" | "ns") {
                let raw: String = self.chars[start..suffix_start].iter().collect();
                let value = raw.parse::<f64>().map_err(|err| {
                    self.error_at(format!("Invalid number {raw}: {err}"), line, col)
                })?;
                return Ok(Token::new(
                    TokenType::FloatLit,
                    TokenValue::Float(to_seconds(value, &suffix)),
                    line,
                    col,
                ));
            }
            self.pos = suffix_start;
        }

        let text: String = self.chars[start..self.pos].iter().collect();
        if is_float || text.contains('.') {
            let value = text
                .parse::<f64>()
                .map_err(|err| self.error_at(format!("Invalid float {text}: {err}"), line, col))?;
            Ok(Token::new(
                TokenType::FloatLit,
                TokenValue::Float(value),
                line,
                col,
            ))
        } else {
            let value = text
                .parse::<i64>()
                .map_err(|err| self.error_at(format!("Invalid int {text}: {err}"), line, col))?;
            Ok(Token::new(
                TokenType::IntLit,
                TokenValue::Int(value),
                line,
                col,
            ))
        }
    }

    fn lex_string(&mut self, line: usize, col: usize) -> Result<Token, LexError> {
        self.advance();
        let mut segments = Vec::new();
        let mut buffer = String::new();

        while self.pos < self.chars.len() {
            let ch = self.peek(0);
            if ch == '"' {
                self.advance();
                if !buffer.is_empty() {
                    segments.push(StringTokenPart::Text(std::mem::take(&mut buffer)));
                }
                return Ok(Token::new(
                    TokenType::String,
                    TokenValue::String(segments),
                    line,
                    col,
                ));
            }
            if ch == '\\' {
                if self.peek(1) == '(' {
                    if !buffer.is_empty() {
                        segments.push(StringTokenPart::Text(std::mem::take(&mut buffer)));
                    }
                    self.advance();
                    self.advance();
                    let (capture, tokens) = self.lex_interp_expr()?;
                    segments.push(StringTokenPart::Interp { capture, tokens });
                    continue;
                }
                self.advance();
                let escaped = self.advance();
                let value = match escaped {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    '"' => '"',
                    '\\' => '\\',
                    other => other,
                };
                buffer.push(value);
                continue;
            }
            buffer.push(self.advance());
        }

        Err(self.error_at("Unterminated string literal".to_string(), line, col))
    }

    fn lex_interp_expr(&mut self) -> Result<(Option<String>, Vec<Token>), LexError> {
        let mut capture = None;
        let mut tokens = Vec::new();
        let mut depth = 1usize;

        let saved_pos = self.pos;
        let saved_line = self.line;
        let saved_col = self.col;
        self.skip_whitespace_and_comments();
        if is_ident_start(self.peek(0)) {
            let start = self.pos;
            while is_ident_continue(self.peek(0)) {
                self.advance();
            }
            let word: String = self.chars[start..self.pos].iter().collect();
            if word == "once" || word == "always" {
                capture = Some(word);
            } else {
                self.pos = saved_pos;
                self.line = saved_line;
                self.col = saved_col;
            }
        } else {
            self.pos = saved_pos;
            self.line = saved_line;
            self.col = saved_col;
        }

        while self.pos < self.chars.len() {
            self.skip_whitespace_and_comments();
            if self.pos >= self.chars.len() {
                break;
            }
            let ch = self.peek(0);
            if ch == '(' {
                depth += 1;
                tokens.push(self.next_token()?);
            } else if ch == ')' {
                depth -= 1;
                if depth == 0 {
                    self.advance();
                    break;
                }
                tokens.push(self.next_token()?);
            } else {
                tokens.push(self.next_token()?);
            }
        }

        Ok((capture, tokens))
    }

    fn skip_whitespace_and_comments(&mut self) {
        loop {
            if self.pos >= self.chars.len() {
                break;
            }
            let ch = self.peek(0);
            if matches!(ch, ' ' | '\t' | '\r' | '\n') {
                self.advance();
            } else if ch == '/' && self.peek(1) == '/' {
                while self.pos < self.chars.len() && self.peek(0) != '\n' {
                    self.advance();
                }
            } else if ch == '/' && self.peek(1) == '*' {
                self.advance();
                self.advance();
                while self.pos < self.chars.len() {
                    if self.peek(0) == '*' && self.peek(1) == '/' {
                        self.advance();
                        self.advance();
                        break;
                    }
                    self.advance();
                }
            } else {
                break;
            }
        }
    }

    fn peek(&self, offset: usize) -> char {
        self.chars.get(self.pos + offset).copied().unwrap_or('\0')
    }

    fn advance(&mut self) -> char {
        let ch = self.peek(0);
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        ch
    }

    fn error_at(&self, message: String, line: usize, col: usize) -> LexError {
        LexError { message, line, col }
    }
}

fn is_ident_start(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

fn is_ident_continue(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_'
}

fn to_seconds(value: f64, suffix: &str) -> f64 {
    match suffix {
        "s" => value,
        "ms" => value * 1e-3,
        "us" => value * 1e-6,
        "ns" => value * 1e-9,
        _ => value,
    }
}

fn format_float(value: f64) -> String {
    let mut text = value.to_string();
    if text.contains('e') || text.contains('E') {
        return text;
    }
    if text.contains('.') {
        while text.ends_with('0') {
            text.pop();
        }
        if text.ends_with('.') {
            text.push('0');
        }
    }
    text
}
